reps/2023-8-18-ray-on-spark-autoscaling.md

```diff
@@ -0,0 +1,204 @@
+above YAML configuration, it configures autoscaler with "spark" `NodeProvider`
+backend. When Ray cluster is requested to scale up,
+the configured `NodeProvider` is called to trigger Ray worker node setup.
+
+This is `NodeProvider` [interfaces](https://github.com/ray-project/ray/blob/e1c48ab8b1225daa44ea63ed1e7dc956b0e7f411/python/ray/autoscaler/node_provider.py#L13).
+
+### How to implement spark node provider?
+
+For each Ray worker node, we start a spark job with only one spark
+task to hold this Ray worker node, and we set a unique spark job group ID to
+this spark job. When we need to terminate this Ray worker node, we cancel
+the corresponding spark job by canceling corresponding spark job group, so that
+the spark job and its spark task are killed, then it triggers the Ray worker node termination.
+
+Because we can only cancel spark job not spark task when we need to scale down a Ray worker node,
+we have to have one spark job for each Ray worker node.
+
+One thing critical is that spark node provider runs in autoscaler process that is different process
+than the one that executes "setup_ray_cluster" API. User calls "setup_ray_cluster" in
+spark application driver node, and the semantic is "setup_ray_cluster" requests spark resources from
+this spark application. Internally, "setup_ray_cluster" should use "spark session" instance
+to request spark application resources. But spark node provider runs in another python process,
+how to share current process spark session to the separate NodeProvider process? The solution is
+setting up a spark job server that runs inside spark application driver process (the process
+that calls "setup_ray_cluster" API), and in NodeProvider process, it sends RPC request to
+the spark job server for creating spark jobs in the spark application. Note that we cannot
+create another spark session in NodeProvider process, because if doing so, it means we create
+another spark application, and then it causes NodeProvider requests resources belonging to
+the new spark application, but we need to ensure all requested spark resources belong to
+the original spark application that calls "setup_ray_cluster" API.
+
+The overall architecture is:
+
+![ray on spark autoscaling architecture](https://github.com/ray-project/ray/assets/19235986/9f1f30bb-e395-4d98-8c8d-5fa60675bafa)
+
+
+### Key questions
+
+
+#### We can use `ray up` command to set up a ray cluster with autoscaling, why we don't call `ray up` command in ray on spark autoscaling implementation ?
+
+In ray on spark, we only provides a python API `setup_ray_cluster` and it does not have a CLI. So in `setup_ray_cluster` implementation, we need to generate autoscale config YAML file according to `setup_ray_cluster` argument values, and then launch the ray head node with "--autoscaling-config" option. In this way ray on spark code can manage ray head node process and ray worker nodes easier.
```

Comment:
So with these 2 constraints, it is, however, still possible to support `ray up` with some sort of config checks and extra works at the config generation routine in ray up right?

Is there gonna be a long term plan to eventually support `ray up` cli?

Reply:
Yes we can consider migrating code to use `ray up` in future, if there are strong reasons and benefits.